homeassistant/components/osoenergy/water_heater.py

```diff
@@ -0,0 +1,194 @@
+"""Support for OSO Energy water heaters."""
+from collections.abc import Mapping
+from typing import Any
+
+from homeassistant.components.water_heater import (
+    WaterHeaterEntity,
+    WaterHeaterEntityFeature,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import STATE_OFF, STATE_ON, UnitOfTemperature
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.device_registry import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+import homeassistant.util.dt as dt_util
+
+from . import OSOEnergyEntity
+from .const import DOMAIN
+
+EXTRA_HEATER_ATTR: dict[str, dict[str, Any]] = {
```

Comment:
I'm not sure this is wanted. We normally recommend exposing these sort of things as additional entities. For example, the `heater_state` should be a binary sensor, the heater mode an enum sensor (or a select if it's possible to set the mode) and so on.